File: packages/agent-runtime/src/attachments.mjs
import { Buffer } from 'node:buffer';

export const MAX_IMAGE_BASE64_LENGTH = 3 * 1024 * 1024;

/** Called before launching a provider, including when the runtime is invoked without Rust. */
export function validateAttachments(attachments, adapter) {
  if (attachments == null) return [];
  if (!Array.isArray(attachments)) throw new Error('Image attachments must be an array');
  if (!attachments.length) return [];
  if (!['codex', 'claude'].includes(adapter))
    throw new Error(
      'Image attachments are supported by Codex and Claude. Choose one of these agents or remove the images.',
    );
  if (attachments.length > 5) throw new Error('Attach up to 5 images per message');
  let total = 0;
  for (const image of attachments) {
    if (
      !image ||
      typeof image.name !== 'string' ||
      !image.name.trim() ||
      image.name.length > 255 ||
      [...image.name].some(
        (character) => character.charCodeAt(0) < 32 || character.charCodeAt(0) === 127,
      )
    )
      throw new Error('Every image must have a valid filename (up to 255 characters)');
    if (!['image/png', 'image/jpeg', 'image/webp', 'image/gif'].includes(image.mime_type))
      throw new Error('Choose PNG, JPEG, WebP or GIF images');
    if (typeof image.data !== 'string' || !image.data)
      throw new Error('Image data is invalid. Attach the file again.');
    total += image.data.length;
    if (total > MAX_IMAGE_BASE64_LENGTH)
      throw new Error('Images must total 2.25 MiB or less. Choose smaller images.');
    if (Buffer.from(image.data, 'base64').toString('base64') !== image.data)
      throw new Error('Image data is invalid. Attach the file again.');
  }
  return attachments;
}

export function codexImageInput(prompt, attachments) {
  return [
    { type: 'text', text: prompt },
    ...validateAttachments(attachments, 'codex').map((image) => ({
      type: 'image',
      url: `data:${image.mime_type};base64,${image.data}`,
    })),
  ];
}

export async function* claudeImageInput(prompt, attachments) {
  // SDKUserMessage.message is Anthropic MessageParam; image source uses native base64 content.
  yield {
    type: 'user',
    parent_tool_use_id: null,
    message: {
      role: 'user',
      content: [
        { type: 'text', text: prompt },
        ...validateAttachments(attachments, 'claude').map((image) => ({
          type: 'image',
          source: { type: 'base64', media_type: image.mime_type, data: image.data },
        })),
      ],
    },
  };
}

File: crates/runhq-core/src/agents/attachments.rs
use super::{invalid, AgentAttachment};
use crate::AppResult;
use base64::{engine::general_purpose::STANDARD, Engine};

const MAX_IMAGE_BASE64_LENGTH: usize = 3 * 1024 * 1024;

/// Validate before reserving a turn or starting a provider so a failed upload is retryable.
pub(super) fn validate_attachments(
    attachments: &[AgentAttachment],
    adapter: &str,
) -> AppResult<()> {
    if attachments.is_empty() {
        return Ok(());
    }
    if !matches!(adapter, "codex" | "claude") {
        return Err(invalid("Image attachments are supported by Codex and Claude. Choose one of these agents or remove the images."));
    }
    if attachments.len() > 5 {
        return Err(invalid("Attach up to 5 images per message"));
    }
    let mut total = 0usize;
    for attachment in attachments {
        if attachment.name.trim().is_empty()
            || attachment.name.chars().count() > 255
            || attachment.name.chars().any(char::is_control)
        {
            return Err(invalid(
                "Every image must have a valid filename (up to 255 characters)",
            ));
        }
        if !matches!(
            attachment.mime_type.as_str(),
            "image/png" | "image/jpeg" | "image/webp" | "image/gif"
        ) {
            return Err(invalid("Choose PNG, JPEG, WebP or GIF images"));
        }
        total = total.saturating_add(attachment.data.len());
        if total > MAX_IMAGE_BASE64_LENGTH {
            return Err(invalid(
                "Images must total 2.25 MiB or less. Choose smaller images.",
            ));
        }
        if attachment.data.is_empty() || STANDARD.decode(&attachment.data).is_err() {
            return Err(invalid("Image data is invalid. Attach the file again."));
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    fn image() -> AgentAttachment {
        AgentAttachment {
            name: "example.png".into(),
            mime_type: "image/png".into(),
            data: "aGVsbG8=".into(),
        }
    }
    #[test]
    fn validates_supported_images_and_keeps_plain_text_compatible() {
        assert!(validate_attachments(&[], "acp").is_ok());
        assert!(validate_attachments(&[image()], "codex").is_ok());
        assert!(validate_attachments(&[image()], "claude").is_ok());
        assert!(validate_attachments(&[image()], "opencode").is_err());
        assert!(validate_attachments(&[image()], "acp").is_err());
        assert!(validate_attachments(&[image()], "terminal").is_err());
    }
    #[test]
    fn bounds_images_and_rejects_invalid_content_before_provider_start() {
        assert!(validate_attachments(&vec![image(); 6], "codex").is_err());
        let mut invalid_image = image();
        invalid_image.mime_type = "application/javascript".into();
        assert!(validate_attachments(&[invalid_image], "codex").is_err());
        let mut invalid_image = image();
        invalid_image.data = "not base64".into();
        assert!(validate_attachments(&[invalid_image], "claude").is_err());
        let mut large = image();
        large.data = "A".repeat(MAX_IMAGE_BASE64_LENGTH);
        assert!(validate_attachments(&[large, image()], "codex").is_err());
    }
    #[test]
    fn deserializes_old_plain_text_turns_without_attachments() {
        let input: super::super::AgentTurnInput = serde_json::from_value(serde_json::json!({
            "session_id":"s", "request_id":"r", "prompt":"hello"
        }))
        .unwrap();
        assert!(input.attachments.is_empty());
    }
}
